### PR TITLE
[updated] Incorporate Cajita / Cabana::Grid load balancer

### PR DESCRIPTION
**Pull request by cz4rs**
Updated version of  #104. I haven't touched the logic of it all, just a rebase, clean-up and dropping some redundant changes.

- resolved all conflicts and fixed deprecation warnings
- added `--sparse` command line flag for an unbalanced system to avoid having to change the source
- fixed data files naming scheme to facilitate processing the output with ParaView

### CODE REVIEW

**Comment by cz4rs** [comment]
@streeve @aetx This is an updated version of PR #104 (see PR description for an overview of changes). The CI is pending approval, but based on the runs in my fork (https://github.com/cz4rs/CabanaMD/pull/1), all builds are passing.
Was there any reason that #104 was not merged?

For the context, we are exploring using CabanaMD as a benchmark for load balancing with communication.

**Comment by aetx** [comment]
If I remember correctly, the only ugly hack was in `cabanamd_impl.h` to create an unbalanced load. By default, the whole system is uniformly filled (see https://github.com/ECP-copa/CabanaMD/pull/104#issuecomment-1030589568).

I don't remember another reason for it being not merged.

**Comment by streeve** [comment]
Sorry I lost track of this PR and thanks very much for pushing this forward. Would you be okay with merging your general (and very helpful) changes in one PR, then merging the load balancing in a second?

**Comment by cz4rs** [comment]
Sure, I will split the changes.
Please note that I'm on vacation now, so this will need to wait a bit (I'll
be back on September 9th).

pt., 30 sie 2024, 16:12 użytkownik Sam Reeve ***@***.***>
napisał:

**Comment by cz4rs** [comment]
@streeve Please have a look.
This is pretty much just a cleaned up version of #104. I have rebased this on top of `master` and cleaned up my commits a bit.

I have some more changes (extending LAMMPS commands support) in https://github.com/DARMA-tasking/CabanaMD/pull/3, I will post those in a separate PR.

**Comment by cz4rs** [comment]
@streeve Ping

File: src/inputCL.cpp [hunk 1]
```diff
@@ -116,6 +116,8 @@ void InputCL::read_args( int argc, char *argv[] )
                  "                                (N = positive integer)\n",
                  "                                (PATH = location of ",
                  "directory)\n" );
+            log( std::cout, "  --sparse : "
```
Comment (streeve) [inline src/inputCL.cpp]: I would prefer `--vacuum`, but not a big deal either way

File: src/inputFile_impl.h [hunk 2]
```diff
@@ -455,7 +512,14 @@ void InputFile<t_System>::create_lattice( Comm<t_System> *comm )
     T_X_FLOAT max_z = lattice_constant * lattice_nz;
     std::array<T_X_FLOAT, 3> global_low = { 0.0, 0.0, 0.0 };
     std::array<T_X_FLOAT, 3> global_high = { max_x, max_y, max_z };
-    system->create_domain( global_low, global_high );
+    if ( commandline.sparse )
```
Comment (streeve) [inline src/inputFile_impl.h]: Since this commandline arg is already set up, can we input the multiplier here instead of assuming 2?

**Review by streeve — COMMENTED** [review]
Thanks for the reminder. Do you mind squashing @aetx's changes into a single "add Cabana load balancing" commit? Otherwise a couple very minor points and we can (finally) get this in

File: src/inputFile_impl.h (hunk 2 above)
Comment (cz4rs) [reply src/inputFile_impl.h]: Sure, I have missed this somehow :man_facepalming:

Comment (streeve) [reply src/inputFile_impl.h]: I think that was the original behavior to get results, but this will be much more general

Comment (cz4rs) [reply src/inputFile_impl.h]: Done. Example below with `1.5` multiplier (`mpiexec -n 8 ./build/bin/cbnMD -il input/in.lj --vacuum 1.5`).

image

File: src/inputCL.cpp (hunk 1 above)
Comment (cz4rs) [reply src/inputCL.cpp]: Renamed to `--vacuum`.

**Review by streeve — APPROVED** [review]
Looks good - thanks one more time to @aetx for the original work and @cz4rs for getting this over the finish line

**Pull request merged**